### DISCUSSION TITLE
Calling a function from a string

### DISCUSSION THREAD

**Discussion by CrimboJimbo** [cc-tweaked/CC-Tweaked#1979] [Q&A] [answered]
I have a file uploaded to github that I am trying to call a function out of. The format is: `[Key] {Function}`. I search through the document to find the correct key then pull the function name out of the `{}` but is refuses to call the function with that name. I have been looking online for was to do this with things like, `_G[c]()` but I have had no luck.

My current code is as follows:

```write(">")
local input = read() --Replace with input from chat
local request = http.get("https://raw.githubusercontent.com/CrimboJimbo/MinecraftDadBot/refs/heads/main/Dad%20Commands.txt")
local inputMod = string.lower(input)
local txt = request.readAll()
local a,b,c
-- print(txt)
function TestFunc()
    print('Yippie!')
end

print(inputMod)
a,b = string.find(txt,inputMod)
if a == nil or b == nil then
    error("idiot")
else
    a,b = string.find(txt,"%b{}",b+1)
    print(a,b)
    c = string.sub(txt,a+1,b-1)
    -- local scrum = load("return function() "..c.." end")
    -- local ok, scrim = pcall(scrum)
    -- print(ok)
    -- _G.scrim()
    _G[c]()
end
request.close()
```

The print output show that `c = TestFunc()` but nothing works when trying to call the function. I've also tried to remove the `()` inside the `c` string but I get the same error messages. any ideas?

Just for context I'm working on an in-game chat bot called DadBot. The path for the TxT I'm using is: `https://raw.githubusercontent.com/CrimboJimbo/MinecraftDadBot/refs/heads/main/Dad%20Commands.txt`
![Code_mhJUF2Ja1i](https://github.com/user-attachments/assets/087b3f09-94d7-4a80-b769-3caae8f6d01d)

**Comment by zyxkad** [comment]
`local scrim` defines a local variable, which not exists in `_G`. Therefore you cannot do `_G.scrim()`, but just invoke `scrim()` is what you want

**Reply by CrimboJimbo** [reply]
That still gives an error sadly.
```
if a == nil or b == nil then
    error("idiot")
else
    a,b = string.find(txt,"%b{}",b+1)
    print(a,b)
    c = string.sub(txt,a+1,b-1)
    local scrum = load("return function() "..c.." end")
    local ok, scrim = pcall(scrum)
    print(ok)
    scrim()
    -- _G[c]()
end
```
![Code_wZ2PKeG0pL](https://github.com/user-attachments/assets/9e42a1dd-4333-434d-8e8f-3933c1560e2c)

**Reply by zyxkad** [reply]
This is be cause the environment is empty, you need pass `_G` as the fourth argument to make the `load` function use the global environment

**Reply by CrimboJimbo** [reply]
I'm not sure I understand. I don't really know what the `load` function is doing to be honest, most of the descriptions online didn't make much sense to me. What should I pass as the 2nd and 3rd arguments to load?

**Reply by zyxkad** [reply]
https://www.lua.org/manual/5.3/manual.html#pdf-load

Just pass `nil` for the second and third value to have the default behaviour.

**Reply by CrimboJimbo** [reply]
so I'm using `load("return function() "..c.." end",nil,"bt",_G)` now but still getting the same error. I read all the documentation on the load function and how lua processes chunks but I can't seem to get it to run the TestFunc from the string.

**Reply by zyxkad** [reply]
Have you defined the `TestFunc` as a global function?

**Reply by CrimboJimbo** [reply]
Do you do that differently than just:
```
function TestFunc()
    print('Yippie!')
end
```
That is what I have at the top of my code

**Reply by zyxkad** [reply]
Can you re-paste the latest codes and the output again?

**Reply by CrimboJimbo** [reply]
Edit: Sorry, I forgot to revert some testing I was doing, this is the code that shows the most promise currently

```
write(">")
local input = read() --Replace with input from chat
local request = http.get("https://raw.githubusercontent.com/CrimboJimbo/MinecraftDadBot/refs/heads/main/Dad%20Commands.txt")
local inputMod = string.lower(input)
local txt = request.readAll()
local a,b,c
-- print(txt)
function TestFunc()
    print('Yippie!')
end

print(inputMod)
a,b = string.find(txt,inputMod)
if a == nil or b == nil then
    error("idiot")
else
    a,b = string.find(txt,"%b{}",b+1)
    print(a,b)
    c = string.sub(txt,a+1,b-1)
    Scrum = load("return function() "..c.." end",nil,"bt",_G)
    Ok, Scrim = pcall(Scrum)
    print(Ok)
    Scrim()
    -- _G[c]()
end
request.close()
```
![Code_ByqaSjYOMN](https://github.com/user-attachments/assets/5aeabc8e-ed30-4a40-8280-ff703bbd454e)

**Reply by fatboychummy** [reply]
CC's `_ENV` and `_G` are different tables, unlike in base lua. You want to pass `_ENV` instead of `_G`.

**Reply by fatboychummy** [reply, marked as answer]
To clarify, when you declare a variable (or function) without the `local` keyword, Lua always puts the value into `_ENV`. This is the case in CC as well, but CC creates an `_ENV` that is specific to each program, then an actual global `_G` (whereas in base lua, both tables are the same).

If you want something put in `_G` instead of `_ENV`, you need to prefix your declaration with `_G.`, i.e: `_G.variable = "global variable"`

**Reply by CrimboJimbo** [reply]
That was it! Thank you!
